graph/patterns/snapshots.md

```diff
@@ -0,0 +1,164 @@
+# Snapshots
+
+Microsoft Graph API Design Pattern
+
+*The snapshots pattern allows access to data for an individual resource even when that resource no longer exists, while still allowing to easily correlate references to that resource.*
+
+## Problem
+
+There are often times when data on an entity (usually a subset) needs to be exposed even after the entity has been removed or updated. 
+This data is representating that entity during a snapshot in time. 
+However, simply duplicating the data prevents correlating that snapshot data with the entity itself or with other data that are also snapshots in time.
+
+An example would be a log of `user`s who have signed in.
+The log could have an entry for when the `user` signed in, as well as details about the user.
+But without a way to correlate this signin data to the `user` itself, further investigations becomes difficult.
+
+## Solution
+
+The solution is to have a complex type which represents the resource as a snapshot in time, while also providing a navigation property to the resource itself.
+The following CSDL demonstrates this with the `user` signin example:
+
+```xml
+<EntityType Name="user">
+  <Key>
+    <PropertyRef Name="id" />
+  </Key>
+  <Property Name="id" Type="Edm.String" Nullable="false" />
+  <Property Name="displayName" Type="Edm.String" Nullable="false" />
+  <Property Name="userPrincipalName" Type="Edm.String" Nullable="false" />
+  <!--
+  Other existing propeties go here.
+  -->
+</EntityType>
+
+<EntityType Name="userSignIn">
+  <Key>
+    <PropertyRef Name="id" />
+  </Key>
+  <Property Name="id" Type="Edm.String" Nullable="false" />
+  <Property Name="attemptDateTime" Type="Edm.DateTimeOffset" Nullable="false" />
+  <Property Name="requestedPermissions" Type="Collection(Edm.String)" Nullable="false" />
+  <Property Name="userSnapshot" Type="self.userSnapshot" Nullable="false" />
+</EntityType>
+
+<ComplexType Name="userSnapshot">
+  <Property Name="id" Type="Edm.String" Nullable="false" />
+  <Property Name="displayName" Type="Edm.String" Nullable="false" />
+  <Property Name="userPrincipalName" Type="Edm.String" Nullable="false" />
+  <!--
+  For this example, only these properties are being stored in the snapshot. Any subset of properties from `user` is valid to store in a snapshot type. The properties should have the same names, types, nullability, etc. as the entity that is referred to.
+  -->
+  <NavigationProperty Name="user" Type="self.user" Nullable="true" /> <!--This is nullable because the user may no longer exist.-->
+</ComplexType>
+```
+
+TODO example with nested snapshots?
+
+## When to use this pattern
```

Comment:
From discussion with Tyler, please make it clear that only the *relevant* properties need to be on the snapshot (maybe have a comparison to a "non-snapshot" version of the same API).